internal/testutils/golden.go

```diff
@@ -8,43 +8,131 @@ import (
 	"testing"
 
 	"github.com/stretchr/testify/require"
+	"gopkg.in/yaml.v2"
 )
 
 var update bool
 
-// LoadWithUpdateFromGolden loads the element from a plaintext golden file in testdata/golden.
-// It will update the file if the update flag is used prior to deserializing it.
-func LoadWithUpdateFromGolden(t *testing.T, data string) string {
+type goldenOptions struct {
+	goldenPath string
+}
+
+// GoldenOption is a supported option reference to change the golden files comparison.
+type GoldenOption func(*goldenOptions)
+
+// WithGoldenPath overrides the default path for golden files used.
+func WithGoldenPath(path string) GoldenOption {
+	return func(o *goldenOptions) {
+		if path != "" {
+			o.goldenPath = path
+		}
+	}
+}
+
+// LoadWithUpdateFromGolden loads the element from a plaintext golden file.
+// It will update the file if the update flag is used prior to loading it.
+func LoadWithUpdateFromGolden(t *testing.T, data string, opts ...GoldenOption) string {
 	t.Helper()
 
-	goldPath := filepath.Join("testdata", "golden", NormalizeGoldenName(t, t.Name()))
+	o := goldenOptions{
+		goldenPath: GoldenPath(t),
+	}
+
+	for _, opt := range opts {
+		opt(&o)
+	}
 
 	if update {
-		t.Logf("updating golden file %s", goldPath)
-		err := os.MkdirAll(filepath.Dir(goldPath), 0750)
+		t.Logf("updating golden file %s", o.goldenPath)
+		err := os.MkdirAll(filepath.Dir(o.goldenPath), 0750)
 		require.NoError(t, err, "Cannot create directory for updating golden files")
-		err = os.WriteFile(goldPath, []byte(data), 0600)
+		err = os.WriteFile(o.goldenPath, []byte(data), 0600)
 		require.NoError(t, err, "Cannot write golden file")
 	}
 
-	want, err := os.ReadFile(goldPath)
+	want, err := os.ReadFile(o.goldenPath)
 	require.NoError(t, err, "Cannot load golden file")
 
 	return string(want)
 }
 
+// LoadGenericWithUpdateFromGolden load the generic element from a serialized golden file.
+// It will update the file if the update flag is used prior to deserializing it.
+func LoadGenericWithUpdateFromGolden[E any](t *testing.T, got E, opts ...GoldenOption) E {
```

Comment:
This is a weird state, I agree. The thing is: as we discussed on MM, the idea was to keep the LoadFromGolden dealing in strings only, but later down the line (in the admxgen package) there was a whole parallel set of golden functionalities implemented inside the tests and I decided to move them to testutils, where they belong.
Also, despite having the logic kind of implemented twice, I kind of like having it for two reasons:
- Serialization with strings do work, but will break the ones that have color tags, so that's a no no. However, we really benefit from having a generic one that is able to serialize structs, slices, etc. (specially in the long term).
- Having both allows us to use the string one for strings, which is faster and used by a lot of our tests, and keep the generics one for more complex datatypes that would require serialization to be written down.

But I totally agree the name is weird and that I should figure out a way of stopping it from being called to strings.

Reply:
I’ll give the refactoring a go after this PR I think to reuse the string one. I’m happy to give it a try. Agreed that we should change the name anyway in that PR.

Reply:
I could use reflect to try and check if the provided E needs serialization, wdyt?

Reply:
I was thinking about the other way around, I’ll try this afternoon to show you what I meant

Reply:
> I could use reflect to try and check if the provided E needs serialization, wdyt?

Actually I would restrict to what the yaml marhsall/unmarshall use. As it’s interface{} for now, let’s not bother in the end. The other changes looks good!
I’m sad we can’t override the same function (one for string, the other for "the rest") and keep a single function name, but this is not doable (which is understandable: what happens if you have 2 functions using generic where the constraints could work with this type?